src/Microsoft.TemplateEngine.Edge/Settings/ComponentManager.cs

```diff
@@ -19,25 +19,46 @@ internal class ComponentManager : IComponentManager
         private readonly Dictionary<Type, HashSet<Guid>> _componentIdsByType;
         private readonly SettingsStore _settings;
         private readonly ISettingsLoader _loader;
+        private readonly ComponentCache _componentCache = new ComponentCache();
 
-        private interface ICache
+        private class ComponentCache 
         {
-            void AddPart(IIdentifiedComponent component);
-        }
+            private readonly Dictionary<Type, Dictionary<Guid, object>> _parts = new Dictionary<Type, Dictionary<Guid, object>>();
 
-        private class Cache<T> : ICache
-            where T : IIdentifiedComponent
-        {
-            public static readonly Cache<T> Instance = new Cache<T>();
+            public bool TryGetPart<T>(Guid id, out T component) where T : IIdentifiedComponent
+            {
+                component = default;
+                if (!_parts.TryGetValue(typeof(T), out Dictionary<Guid, object> typeCache) || typeCache == null || !typeCache.ContainsKey(id))
```

Comment:
If we really want to complicate things, last `ContainsKey` could be `TryGetValue` :)